include/mxnet/runtime/packed_func.h

```diff
@@ -0,0 +1,1254 @@
+ * under the License.
+ */
+
+/*!
+ * \file runtime/packed_func.h
+ * \brief Adapted from incubator-tvm/include/tvm/runtime/packed_func.h
+ * Type-erased function used across MXNET API.
+ */
+#ifndef MXNET_RUNTIME_PACKED_FUNC_H_
+#define MXNET_RUNTIME_PACKED_FUNC_H_
+
+#include <dmlc/logging.h>
+#include <mxnet/runtime/c_runtime_api.h>
+#include <mxnet/runtime/object.h>
+#include <mxnet/runtime/ndarray.h>
+#include <mxnet/runtime/container.h>
+#include <mxnet/runtime/ffi_helper.h>
+#include <mxnet/runtime/data_type.h>
+#include <mxnet/node/container.h>
+#include <mxnet/ir/expr.h>
+#include <mxnet/ndarray.h>
+#include <mxnet/base.h>
+#include <functional>
+#include <tuple>
+#include <vector>
+#include <string>
+#include <limits>
+#include <memory>
+#include <utility>
+#include <type_traits>
+#include <sstream>
+
+namespace mxnet {
+// forward declarations
+// class Integer;
+// class Expr;
+
+namespace runtime {
+
+/*!
+ * \brief Runtime utility for getting custom type name from code
+ * \param type_code Custom type code
+ * \return Custom type name
+ */
+// MXNET_DLL std::string GetCustomTypeName(uint8_t type_code);
+
+/*!
+ * \brief Runtime utility for checking whether custom type is registered
+ * \param type_code Custom type code
+ * \return Bool representing whether type is registered
+ */
+// MXNET_DLL bool GetCustomTypeRegistered(uint8_t type_code);
+
+/*!
+ * \brief Runtime utility for parsing string of the form "custom[<typename>]"
+ * \param s String to parse
+ * \param scan pointer to parsing pointer, which is scanning across s
+ * \return type code of custom type parsed
+ */
+// MXNET_DLL uint8_t ParseCustomDatatype(const std::string& s, const char** scan);
+
+/*!
+ * \brief convert a string to TVM type.
+ * \param s The string to be converted.
+ * \return The corresponding tvm type.
+ */
+inline DLDataType String2DLDataType(std::string s);
+
+// forward declarations
+class MXNetArgs;
+class MXNetArgValue;
+class MXNetRetValue;
+class MXNetArgsSetter;
+
+/*!
+ * \brief Packed function is a type-erased function.
+ *  The arguments are passed by packed format.
+ *
+ *  This is an useful unified interface to call generated functions,
+ *  It is the unified function function type of TVM.
+ *  It corresponds to TVMFunctionHandle in C runtime API.
+ */
+class PackedFunc {
+ public:
+  /*!
+   * \brief The internal std::function
+   * \param args The arguments to the function.
+   * \param rv The return value.
+   *
+   * \code
+   *   // Example code on how to implemented FType
+   *   void MyPackedFunc(MXNetArgs args, MXNetRetValue* rv) {
+   *     // automatically convert arguments to desired type.
+   *     int a0 = args[0];
+   *     float a1 = args[1];
+   *     ...
+   *     // automatically assign values to rv
+   *     std::string my_return_value = "x";
+   *     *rv = my_return_value;
+   *   }
+   * \endcode
+   */
+  using FType = std::function<void (MXNetArgs args, MXNetRetValue* rv)>;
```

Comment:
I have a question about the ABI compatibility of std::functions.

std::functions is a STL container, which has different implementation between different versions of compiler (e.g. gcc4 and gcc5) or different compilers (gcc and clang).

For example, MXNet is built in gcc4, which uses the reference & to store the function arguments. However, the other application is built in gcc5, which uses the rvalue reference && to store the function arguments. The other application could not call the MXNet API by PackedFunc, since the implementations of std::function are different between gcc4 and gcc5.

Reply:
Thanks @wkcn :). Basically `PackedFunc` does not go across the dll boundary. It remains in the MXNet side.

When application A calls PackedFunc, it first passes the function name to MXNet. Then MXNet finds the corresponding PackedFunc, and returns the address of the PackedFunc back to A (A may cache the name-address map). Finally, A invokes that PackedFunc with the address.

To conclude, A is not aware of std::function. It only gets an address or some other unique identifier of PackedFunc. And the address is just a void*, which is ABI compatible.

Reply:
@hzfan Thank you!

Yes. In the application A, we can call `MXNetFuncCall` to call the PackedFunc, call `MXNetFuncFree` to free the PackedFunc, and call `MXNetFuncCreateFromCFunc` like [TVMFuncCreateFromCFunc](https://github.com/apache/incubator-tvm/blob/a43e326fb0250a46c61f726a4633633c2af2bf03/include/tvm/runtime/c_runtime_api.h#L340) in TVM to create new PackedFunc.

However, when the implementations of `std::function` are different between MXNet and the application A, A could not call PackedFunc directly, namely
```c++
PackedFunc *func = GetAPackedFuncFromMXNet();
(*func)(a, b, c);
```

Besides, A could not pass its PackedFunc to MXNet directly, namely
```c++
PackedFunc func = GetAPackedFuncFromA();
MXNetFuncCall(&func, xxx);
```

Reply:
@wkcn Thanks for bringing this up. Yes, I agree that the above two cases trigger unexpected errors.

In the first case, PackedFunc is exposed to A, which is not intended. Also, in this pr, application A is basically cython. As can be seen from `function.pxi` and `convert.pxi`, PackedFunc is not exposed to cython. So the first case does not occur in this pr.

As for the second case, currently this FFI does not support receiving callback functions from outside of MXNet (`MXNetFuncCreateFromCFunc` does not exist). So the second case does not occur either.

Generally in this pr PackedFunc is called through the intended `MXNetFuncCall`, which ensures the ABI-compatibility.

Reply:
@hzfan I see. Thank you for the detailed explaination : )